pxr/usdImaging/usdImaging/delegate.cpp

```diff
@@ -994,6 +1006,29 @@ UsdImagingDelegate::ApplyPendingUpdates()
                          });
         _usdPathsToResync.clear();
 
+        // Sort and find out all the key usd paths for `_GatherDependencies()`
+        SdfPathSet sortedUsdPathsToResync;
+        for (SdfPath const& usdPath: usdPathsToResync) {
+            if (usdPath.IsPropertyPath()) {
+                sortedUsdPathsToResync.emplace(usdPath.GetPrimPath());
+            } else if (usdPath.IsTargetPath()) {
+                sortedUsdPathsToResync.emplace(usdPath.GetParentPath());
+            } else if (usdPath.IsAbsoluteRootOrPrimPath()) {
+                sortedUsdPathsToResync.emplace(usdPath);
+            }
+        }
+
+        // Pre-cache dependencies in parallel
+        WorkDispatcher resyncPathsCacheDispatcher;
+        for (const auto& usdPath: sortedUsdPathsToResync) {
+            auto& affectedCachePaths = _dependenciesCacheMap[usdPath];
```

Comment:
Building "sortedUsdPathsToResync" will be expensive.  You can delete the loop above by running through "usdPathsToResync", doing the GetPrimPath/GetParentPath/etc, and then replacing this line with:

auto pair = _dependenciesCacheMap.insert(usdPath);
if (pair.second) { continue; /* already inserted */ }
auto& affectedCachePaths = pair->first.second;

Reply:
Note also that you might want to add caching for "usdPathsToUpdate" as well.  The paths to resync are a result of USD composition/type changes, whereas updates reflect property value changes; both call _GatherDependencies.